ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY when using local package as dependency

### Last pnpm version that worked:
7.30.5

### pnpm version:
8.1.0

### Code to reproduce the issue:
I uploaded a repro repository for this here: https://github.com/MaSch0212/pnpm-local-package-issue

### Expected behavior:
1. Add a dependency to a folder, e.g.: `"@types/bpmn-js": "file:./@types/bpmn-js"`
2. Run `pnpm install`
   - It runs without issues.
3. Run `pnpm install --frozen-lockfile`
   - Should run without issues.

### Actual behavior:
1. Add a dependency to a folder, e.g.: `"@types/bpmn-js": "file:./@types/bpmn-js"`
2. Run `pnpm install`
   - It runs without issues.
3. Run `pnpm install --frozen-lockfile`
   - Shows `ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY` error

Running any other `pnpm install` command afterwards produces the `ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY` error. In this example I run `pnpm install --frozen-lockfile` to simulate a build pipeline:
![image](https://user-images.githubusercontent.com/8530764/229038774-558d3de6-9b3b-4d27-9cde-1247a817af1a.png)

After running `pnpm install` once, you will always get the error from that point on (unless the local package dependency is removed):
![image](https://user-images.githubusercontent.com/8530764/229037792-28f6e005-26ca-49bf-b05d-b825cf1b7a0c.png)

### Additional information:

 - `node -v` prints: v16.18.1
 - Windows, macOS, or Linux?: Windows

--- a/.changeset/healthy-goats-tell.md
+++ b/.changeset/healthy-goats-tell.md
@@ -1,4 +1,5 @@
 ---
+"@pnpm/dependency-path": patch
 "@pnpm/lockfile-file": patch
 "pnpm": patch
 ---

--- a/.changeset/short-mirrors-exercise.md
+++ b/.changeset/short-mirrors-exercise.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/dependency-path": minor
+---
+
+Export `indexOfPeersSuffix`.

--- a/.changeset/wise-actors-fix.md
+++ b/.changeset/wise-actors-fix.md
@@ -3,4 +3,4 @@
 "pnpm": patch
 ---
 
-Installation should no fail when there is a local dependency that starts in a directory that starts with the `@` char [#6332](https://github.com/pnpm/pnpm/issues/6332).
+Installation should not fail when there is a local dependency that starts in a directory that starts with the `@` char [#6332](https://github.com/pnpm/pnpm/issues/6332).

--- a/.changeset/wise-actors-fix.md
+++ b/.changeset/wise-actors-fix.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/lockfile-file": patch
+"pnpm": patch
+---
+
+Installation should no fail when there is a local dependency that starts in a directory that starts with the `@` char [#6332](https://github.com/pnpm/pnpm/issues/6332).

--- a/lockfile/lockfile-file/src/experiments/inlineSpecifiersLockfileConverters.ts
+++ b/lockfile/lockfile-file/src/experiments/inlineSpecifiersLockfileConverters.ts
@@ -162,7 +162,7 @@ export function revertFromInlineSpecifiersFormat (lockfile: InlineSpecifiersLock
 const PEERS_SUFFIX_REGEX = /(\([^)]+\))+$/
 
 export function convertLockfileV6DepPathToV5DepPath (newDepPath: string) {
-  if (!newDepPath.includes('@', 2)) return newDepPath
+  if (!newDepPath.includes('@', 2) || newDepPath.startsWith('file:')) return newDepPath
   const index = newDepPath.indexOf('@', newDepPath.indexOf('/@') + 2)
   if (newDepPath.includes('(') && index > newDepPath.search(PEERS_SUFFIX_REGEX)) return newDepPath
   return `${newDepPath.substring(0, index)}/${newDepPath.substring(index + 1)}`

--- a/lockfile/lockfile-file/src/experiments/inlineSpecifiersLockfileConverters.ts
+++ b/lockfile/lockfile-file/src/experiments/inlineSpecifiersLockfileConverters.ts
@@ -159,12 +159,10 @@ export function revertFromInlineSpecifiersFormat (lockfile: InlineSpecifiersLock
   return newLockfile
 }
 
-const PEERS_SUFFIX_REGEX = /(\([^)]+\))+$/
-
 export function convertLockfileV6DepPathToV5DepPath (newDepPath: string) {
   if (!newDepPath.includes('@', 2) || newDepPath.startsWith('file:')) return newDepPath
   const index = newDepPath.indexOf('@', newDepPath.indexOf('/@') + 2)
-  if (newDepPath.includes('(') && index > newDepPath.search(PEERS_SUFFIX_REGEX)) return newDepPath
+  if (newDepPath.includes('(') && index > dp.indexOfPeersSuffix(newDepPath)) return newDepPath
   return `${newDepPath.substring(0, index)}/${newDepPath.substring(index + 1)}`
 }
 

--- a/packages/dependency-path/src/index.ts
+++ b/packages/dependency-path/src/index.ts
@@ -26,13 +26,15 @@ export function resolve (
   return resolutionLocation
 }
 
+const PEERS_SUFFIX_REGEX = /(\([^)]+\))+$/
+
 export function tryGetPackageId (registries: Registries, relDepPath: string) {
   if (relDepPath[0] !== '/') {
     return null
   }
   const sepIndex = relDepPath.indexOf('(')
-  if (sepIndex !== -1) {
-    return resolve(registries, relDepPath.slice(0, sepIndex))
+  if (sepIndex !== -1 && relDepPath.search(PEERS_SUFFIX_REGEX) !== -1) {
+    return resolve(registries, relDepPath.replace(PEERS_SUFFIX_REGEX, ''))
   }
   const underscoreIndex = relDepPath.indexOf('_', relDepPath.lastIndexOf('/'))
   if (underscoreIndex !== -1) {

--- a/packages/dependency-path/src/index.ts
+++ b/packages/dependency-path/src/index.ts
@@ -26,15 +26,29 @@ export function resolve (
   return resolutionLocation
 }
 
-const PEERS_SUFFIX_REGEX = /(\([^)]+\))+$/
+export function indexOfPeersSuffix (depPath: string) {
+  if (!depPath.endsWith(')')) return -1
+  let open = true
+  for (let i = depPath.length - 2; i >= 0; i--) {
+    if (depPath[i] === '(') {
+      open = false
+    } else if (depPath[i] === ')') {
+      if (open) return -1
+      open = true
+    } else if (!open) {
+      return i + 1
+    }
+  }
+  return -1
+}
 
 export function tryGetPackageId (registries: Registries, relDepPath: string) {
   if (relDepPath[0] !== '/') {
     return null
   }
-  const sepIndex = relDepPath.indexOf('(')
-  if (sepIndex !== -1 && relDepPath.search(PEERS_SUFFIX_REGEX) !== -1) {
-    return resolve(registries, relDepPath.replace(PEERS_SUFFIX_REGEX, ''))
+  const sepIndex = indexOfPeersSuffix(relDepPath)
+  if (sepIndex !== -1) {
+    return resolve(registries, relDepPath.substring(0, sepIndex))
   }
   const underscoreIndex = relDepPath.indexOf('_', relDepPath.lastIndexOf('/'))
   if (underscoreIndex !== -1) {

--- a/packages/dependency-path/test/index.ts
+++ b/packages/dependency-path/test/index.ts
@@ -180,4 +180,5 @@ test('depPathToFilename()', () => {
 test('tryGetPackageId', () => {
   expect(tryGetPackageId({ default: 'https://registry.npmjs.org/' }, '/foo/1.0.0_@types+babel__core@7.1.14')).toEqual('registry.npmjs.org/foo/1.0.0')
   expect(tryGetPackageId({ default: 'https://registry.npmjs.org/' }, '/foo/1.0.0(@types/babel__core@7.1.14)')).toEqual('registry.npmjs.org/foo/1.0.0')
+  expect(tryGetPackageId({ default: 'https://registry.npmjs.org/' }, '/@(-.-)/foo/1.0.0(@types/babel__core@7.1.14)')).toEqual('registry.npmjs.org/@(-.-)/foo/1.0.0')
 })

